teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+   [images](https://github.com/tektoncd/plumbing/tree/main/tekton/images). Each directory contains a Dockerfile -
+   in a Pipeline that builds all images in this repo, we would want to: 1) determine which Dockerfiles need to be built
+   and 2) for each of those Dockerfiles, build the image. Looping support would allow us to use a Task which builds
+   just one image (e.g. [the kaniko task](https://github.com/tektoncd/catalog/tree/main/task/kaniko/0.4)), and run it
+   for each Dockerfile, instead of having to the task to take multiple Dockerfiles as input. This missing piece of this
+   is supporting the Task which determines with Dockerfiles to build and provides them as an array (instead of
+   encoded in a string).
+
+   ```
+   filter-dockerfiles
+        |
+        | result: list of dockerfiles
+        v
+   (for each: filter-dockerfiles result)
+      kaniko build
+   ```
+
+## Requirements
+
+* Must be possible for a result array to be empty
+* Must be possible to use the array results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has an array param
+* Must be possible to use one specific value in an array result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value at the specified index, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+
+## Proposal
+
+Currently we treat the contents of
+[a Task's result file(s)](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results)
+as a string. The file is located at `$(results.resultName.path)` which expands to `/tekton/results/resultName`.
+In this proposal, we add the option to write to a json file instead, available at `$(results.resultName.jsonPath)`
+which expands to `/tekton/results/resultName.json`. The contents of this file would be treated as json and in addition
+to supporting a string value in this file, we sould support json arrays.
+
+This adds some additional complexity in that there are two paths to which you could write results (the string file
+and the json file), but this option has these advantages:
+
+* Completely backward compatible with current use of the results file
+* Easy to specify string content which is itself json without needing to add additional logic to escape it
+  (this is only the case for string results; not for values within an array which would still need to be escaped)
+
+This feature would be considered alpha and would be
+[gated by the alpha flag](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).
+
+### Notes/Caveats
+
+* What if the file contains json we don't support? (e.g. we don't yet support dictionaries, and wouldn't
+  initially support arrays of arrays)
+    * The TaskRun (or Pipeline Task) would fail
```

Comment:
We could at least partially validate this on the creation of the resource too

Reply:
could you explain a bit more? I'm not sure how we'd know until runtime that something was written to the file in an unexpected format

Reply:
Yeah that is what I mean. We can validate it at runtime and potentially fail the pipeline

Reply:
I'm adding a section "Validating JSON results at runtime" to cover this in more detail and moving this list out of the notes/caveats section